tests/core/utilities/test_encoding.py

```diff
@@ -80,3 +97,122 @@ def test_hex_encode_abi_type(abi_type, value, expected):
 
     actual = hex_encode_abi_type(abi_type, value)
     assert actual == expected
+
+
+@pytest.mark.skipif(sys.version_info.major > 2, reason="these test values only valid for py2")
+@given(
+    st.one_of(st.integers(min_value=0), st.booleans()),
+    st.sampled_from((to_bytes, to_hex, to_decimal)),
+)
+def test_hexstr_if_str_passthrough_py2(val, converter):
+    assert hexstr_if_str(converter, val) == converter(val)
+
+
+@pytest.mark.skipif(sys.version_info.major > 2, reason="these test values only valid for py2")
+@given(
+    st.from_regex(HEX_REGEX),
+    st.sampled_from((to_bytes, to_hex, to_decimal)),
+)
+def test_hexstr_if_str_valid_hex_py2(val, converter):
+    if converter is to_decimal and to_bytes(hexstr=val) == b'':
+        with pytest.raises(ValueError):
+            hexstr_if_str(converter, val)
+    else:
+        assert hexstr_if_str(converter, val) == converter(hexstr=val)
+
+
+@pytest.mark.skipif(sys.version_info.major > 2, reason="these test values only valid for py2")
+@given(
+    st.one_of(st.text(), st.binary()),
+    st.sampled_from((to_bytes, to_hex, to_decimal)),
+)
+def test_hexstr_if_str_invalid_hex_py2(val, converter):
+    try:
+        is_hexstr = (is_hex(val) or val == '')
+    except ValueError:
+        is_hexstr = False
+
+    if not is_hexstr:
+        with pytest.raises(ValueError):
+            hexstr_if_str(converter, val)
+
+
+@pytest.mark.skipif(sys.version_info.major < 3, reason="these test values only valid for py3")
+@given(st.one_of(st.integers(), st.booleans(), st.binary()))
+@example(b'')
+def test_hexstr_if_str_passthrough(val):
+    to_type = Mock(return_value='zoot')
+    assert hexstr_if_str(to_type, val) == 'zoot'
+    assert to_type.call_args == ((val, ), {'hexstr': None})
+
+
+@pytest.mark.skipif(sys.version_info.major < 3, reason="these test values only valid for py3")
+@given(st.from_regex(HEX_REGEX))
+@example('0x')
+@example('0')
+def test_hexstr_if_str_on_valid_hex(val):
+    to_type = Mock(return_value='zoot')
+    assert hexstr_if_str(to_type, val) == 'zoot'
```

Comment:
Are you sure this is true even for small values of `'zoot'`?  What about as `'zoot'` approaches infinity?  Can you link me to the `'zoot'` documentation so that I can read more about this `'zoot'` thing?

🤡

Reply:
[Docs](https://getyarn.io/yarn-clip/c99f2aa3-9ec4-46fb-8488-b8d6f67bba01)